ports/javascript/mphalport.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2013, 2014 Damien P. George and 2017, 2018 Rami Ali
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "library.h"
+#include "mphalport.h"
+
+void mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) {
+    mp_js_write(str, len);
+}
+
+void mp_hal_delay_ms(mp_uint_t ms) {
+    uint32_t start = mp_hal_ticks_ms();
+    while (mp_hal_ticks_ms() - start < ms) {
+    }
```

Comment:
This is a busy wait, correct?

Reply:
That is correct.  Javascript in most (all?) web browsers does not implement sleep.  There are a couple solutions.  The simplest is to busy-wait, at the expense of CPU time.  If the code is in a web worker, you can launch a synchronous get request against something that will never return a result, with a timeout.  Strictly speaking, you can do the same in the main thread too, but that behaviour is deprecated and may not work everywhere forever.  Then again, sleeping in the main thread is frowned upon anyway.

The most proper solution is to set a timer to resume, then suspend the VM.  Javascript closures make it relatively simple to do, but the rest of the code needs to support it via CPS or similar.